Solver : add options for optim 1 and 2 [ANT-2753]

guilpier-code opened the pull request:
Fixes [ANT-2753](https://gopro-tickets.rte-france.com/browse/ANT-2753).

What is done eventually : 
- [x] Setting and checking solver options is spread in different places, we regroup them in one place, for clarity and cohesion.
- [x] adding command line options `--lp-solver-param-optim-1` and `--lp-solver-param-optim-2`
- [x] checking command line options validity
- [x] Unit testing solver command line options separately (instead of testing them through unit testing **generaldata.ini** parameters)
- [x] making solver take command line options into account
- [x] adding documentation about command line options

pet-mit committed:
Support multiple quadratic solvers

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
UI fix + doc

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
updates

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
fix

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
fix

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
cleanup

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
fix UI build

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
revert submodule

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

pet-mit committed:
some sonar fixes

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
read dual values. fix logs. add PDLP support.

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
add comments

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
tests & some refactoring

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
update doc

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
update cucumber test steps

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
rename test suite

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
print error message

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
test new ortools release

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
add scip support

Signed-off-by: Peter Mitri <peter.mitri@rte-france.com>

pet-mit committed:
apply google's fix proposal

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

pet-mit committed:
fix windows CI

pet-mit committed:
Merge remote-tracking branch 'origin/feature/new_quad_solvers' into feature/new_quad_solvers

pet-mit committed:
review1

pet-mit committed:
review2

pet-mit committed:
format code

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

pet-mit committed:
refactor options and add test

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

pet-mit committed:
format code

pet-mit committed:
Merge remote-tracking branch 'origin/feature/new_quad_solvers' into feature/new_quad_solvers

pet-mit committed:
review

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

pet-mit committed:
add comments for ortools bug

pet-mit committed:
format code

pet-mit committed:
review

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

guilpier-code committed:
add solver options (optim 1 and 2) : add options + renaming a function with a misleading name

pet-mit committed:
fix comment

pet-mit commented on src/solver/misc/options.cpp:
```diff
@@ -98,6 +98,22 @@ std::unique_ptr<Yuni::GetOpt::Parser> CreateParser(Settings& settings, StudyLoad
                 "solver-parameters",
                 "Deprecated, use linear-solver-parameters instead.");
 
+    // --lp-solver-param-optim-1
```
don't forget to add the same information to the CLI doc on the website

guilpier-code replied:
This was done in [this PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2699), based on this one.

flomnes replied:
The mentioned PR has nothing to do with this, it must be a mistake

guilpier-code replied:
This doc was added in [this "overall" PR](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2765) (I checked that the link is correct this time :-))

guilpier-code commented on src/solver/application/application.cpp:
```diff
@@ -229,7 +229,10 @@ void Application::readDataForTheStudy(Data::StudyLoadOptions& options)
+// TODO : this function is too long and has a bad name.
+// TODO : we should split it into (at least) 4 functions.
+// TODO : Naming will be easier.
+void Application::readStudy_makeChecks_and_printThings(Data::StudyLoadOptions& options)
```
We quickly renamed this function with a less misleading name, though imperfect.
Previous name was not good at all. This one is long but more accurate.
As specified in comment, we should split this function.

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

guilpier-code committed:
add solver options (optim 1 and 2) : move set and checks on LP solver in one place.

guilpier-code committed:
[skip ci] add solver options (optim 1 and 2) : small improvements

guilpier-code commented on src/api/API.cpp:
```diff
@@ -70,7 +70,7 @@ SimulationResults APIInternal::execute(
     Settings settings;
     auto& parameters = study_->parameters;
-    parameters.optOptions = optOptions;
+    parameters.optOptions << optOptions;
```
```c++
parameters.optOptions = optOptions;
```
is replaced with : 
```c++
parameters.optOptions << optOptions;
```
I suppose that existing **operator<<** is meant to be used here.
@pet-mit : as you introduced this operator, what do you think ?

pet-mit replied:
ok for `<<`, it seems more coherent with CLI use

flomnes replied:
What does this `operator<<` do specifically ?

From @guilpier-code : this operator was removed

guilpier-code replied:
Please see with @pet-mit, its his.

pet-mit replied:
following @guilpier-code 's proposal on [this](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2574) PR, this operator overrides all fields of the optimization options, except for the logs flag, on which it operates a logical OR (code [here](https://github.com/AntaresSimulatorTeam/Antares_Simulator/blob/f26648161d04f1bbac5b17008570bddea701d44a/src/libs/antares/optimization-options/options.cpp#L24))

guilpier-code replied:
The operator << disappeared eventually.

guilpier-code commented on src/libs/antares/checks/checkLoadedInputData.cpp:
```diff
@@ -20,7 +20,6 @@
 */
 
 #include <antares/checks/checkLoadedInputData.h>
-#include <antares/exception/InitializationError.hpp>
```
```c++
#include <antares/exception/InitializationError.hpp>
```
was never used here. We remove it.

guilpier-code commented on src/solver/application/include/antares/application/application.h:
```diff
@@ -121,14 +121,13 @@ class Application final: public Yuni::IEventObserver<Application, Yuni::Policy::
     bool handleOptions(const Data::StudyLoadOptions& options);
     // Return false if the user requested help, true otherwise
     bool parseCommandLine(Data::StudyLoadOptions& options);
-    void handleParserReturn(Yuni::GetOpt::Parser* parser);
```
```c++
void handleParserReturn(Yuni::GetOpt::Parser* parser);
```
is declared but has never been defined or used. We remove it.

guilpier-code committed:
add solver options (optim 1 and 2) : make a function static

guilpier-code committed:
add solver options (optim 1 and 2) : removing expected failing checks in "generaldata.ini" parameters unit tests.

guilpier-code committed:
add solver options (optim 1 and 2) : adding a check on solver options consistency + post-treat solver options

guilpier-code committed:
add solver options (optim 1 and 2) : set solver options depending on the optimization number.

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

guilpier-code committed:
[skip ci] add solver options (optim 1 and 2) : tiny improvements

guilpier-code committed:
Solver options when MILP : quick improvement

guilpier-code committed:
Solver options when MILP : more improvements

flomnes commented on src/libs/antares/checks/checksOnLPsolver.cpp:
```diff
@@ -0,0 +1,48 @@
+    bool found = std::ranges::find(availableSolversList, solverName) != availableSolversList.end();
+    if (!found)
+    {
+        throw Error::InvalidSolver(solverName, boost::algorithm::join(availableSolversList, ","));
```
there is already a function just for that, no need to add new code (boost::algorithm::join)

```
std::string availableOrToolsSolversString();
```

guilpier-code replied:
The function you point internally retrieves the available solvers. 
But in this scope, this list already exists.
As consequence, client code **checkForSolversExistence(...)** (which is now called **availableLinearSolversString()** or **availableQuadraticSolversString()** by the way, due to @pet-mit work on quadratic solver) would result in retrieving this list of solvers twice.

What I did is replace calls to `availableLinearSolversString()`  with : 
```c++
toString(availableLinearSolversList());
```
It results in having less lines of code and less functions.

flomnes commented on src/solver/optimisation/opt_optimisation_lineaire.cpp:
```diff
@@ -262,6 +262,7 @@ bool OPT_OptimisationLineaire(const OptimizationOptions& options,
         OPT_ExportStructures(problemeHebdo, writer);
     }
 
+    options.linearSolverParameters = options.lpSolverParamOptim1;
```
This is bad. Don't let `OPT_OptimisationLineaire` change it's options. What's more `options` is unique (unique like in "not `options[numSpace]`"), so we'll encounter race conditions in parallel simulations.

One solution is to add methods on this structure
`OptimizationOptions opt1 = options.getOptionsOptim1()`
`OptimizationOptions opt2 = options.getOptionsOptim2()`

But then you need to clarify the role of `OptimizationOptions`, and maybe create a new class to retain the options in general (excluding optimization 1 & 2).


```cpp
// Options for a single optimization
class SingleOptOptions
{
};
class OptimizationOptions
{
SingleOptOptions getOptionsOptim1();
SingleOptOptions getOptionsOptim2();
};

// ...
        return runWeeklyOptimization(options.getOptionsOptim2(),
                                     problemeHebdo,
                                     writer,
                                     DEUXIEME_OPTIMISATION,
                                     simulationObserver);
```
WDYT ?

guilpier-code replied:
I agree on race condition, and on the solution you suggest.
We could even have : 
```c++
class SingleOptOptions
{
    std::string solverName;
    std::string solverParameters;
    bool solverLogs;
};

class OptimizationOptions
{
    SingleOptOptions firstOptimOptions;
    SingleOptOptions secondOptimOptions;
    SingleOptOptions quadraticOtimPtions;
};
```

guilpier-code committed:
add solver options (optim 1 and 2) : starting a change in class about optimization options

guilpier-code committed:
add solver options (optim 1 and 2) : code review - splitting class options into 3

guilpier-code committed:
add solver options (optim 1 and 2) : tiny cleaning

guilpier-code committed:
add solver options (optim 1 and 2) : code review - adapting code of part OPT after splitting options class

guilpier-code committed:
add solver options (optim 1 and 2) : add const where needed

flomnes committed:
Typo

flomnes commented on src/api/API.cpp:
```diff
@@ -35,7 +35,7 @@ namespace Antares::API
   const IStudyLoader& study_loader,
   const std::filesystem::path& output,
-  const Antares::Solver::Optimization::OptimizationOptions& optOptions)
+  const Antares::Solver::Optimization::CmdLineOptimOptions& optOptions)
```
revert

guilpier-code replied:
Done

flomnes commented on src/api/API.cpp:
```diff
@@ -58,7 +58,7 @@ SimulationResults APIInternal::run(
 SimulationResults APIInternal::execute(
   const std::filesystem::path& output,
-  const Antares::Solver::Optimization::OptimizationOptions& optOptions) const
+  const Antares::Solver::Optimization::CmdLineOptimOptions& optOptions) const
```
revert

guilpier-code replied:
Done

flomnes commented on src/api/API.cpp:
```diff
@@ -70,7 +70,7 @@ SimulationResults APIInternal::execute(
     Settings settings;
     auto& parameters = study_->parameters;
-    parameters.optOptions = optOptions;
+    parameters.optOptions.initializeWith(optOptions);
```
revert

guilpier-code replied:
Done

flomnes commented on src/api/include/antares/api/solver.h:
```diff
@@ -37,5 +37,5 @@ namespace Antares::API
   const std::filesystem::path& study_path,
   const std::filesystem::path& output,
-  const Antares::Solver::Optimization::OptimizationOptions& optOptions) noexcept;
+  const Antares::Solver::Optimization::CmdLineOptimOptions& optOptions) noexcept;
```
revert

guilpier-code replied:
Done

flomnes commented on src/solver/optimisation/include/antares/solver/optimisation/opt_fonctions.h:
```diff
@@ -51,7 +53,7 @@ void OPT_InitialiserLesBornesDesVariablesDuProblemeLineaire(PROBLEME_HEBDO*,
 void OPT_InitialiserLesCoutsLineaire(PROBLEME_HEBDO*, const int, const int);
 
-bool OPT_PilotageOptimisationLineaire(const OptimizationOptions& options,
+bool OPT_PilotageOptimisationLineaire(OptimizationOptions& options,
```
const

guilpier-code replied:
Done

flomnes commented on src/solver/optimisation/opt_pilotage_optimisation_quadratique.cpp:
```diff
@@ -52,7 +52,7 @@ bool OPT_PilotageOptimisationQuadratique(const OptimizationOptions& options,
             OPT_InitialiserLesCoutsQuadratiques(problemeHebdo, pdtHebdo);
 
-            result = OPT_AppelDuSolveurQuadratique(options, problemeHebdo->ProblemeAResoudre.get())
+            result = OPT_AppelDuSolveurQuadratique(options.quadraticOptimOptions, problemeHebdo->ProblemeAResoudre.get())
                      && result;
```
Pass `SingleOptim` to this function

guilpier-code replied:
That's already what is done

guilpier-code replied:
Done

flomnes reviewed (commented):
linearSolver is still common (in `OptimizationOptions`, not `SingleOpt`), so this shouldn't break Xpansion

We don't want to use a different solver for optim1 & optim2

guilpier-code committed:
add solver options (optim 1 and 2) : corrections and formatting

guilpier-code committed:
add solver options (optim 1 and 2) : fix after code review

guilpier-code committed:
add solver options (optim 1 and 2) : fix after code review

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/new_quad_solvers' into feature/add-solver-options-for-optim-1-and-2

guilpier-code committed:
[skip ci] add solver options (optim 1 and 2) : clang format

guilpier-code committed:
[skip ci] add solver options (optim 1 and 2) : clarifying unit tests a bit

guilpier-code committed:
add solver options (optim 1 and 2) : cleaning + adding a unit test

guilpier-code committed:
add solver options (optim 1 and 2) : add unit tests

guilpier-code committed:
add solver options (optim 1 and 2) : introducing a fixture for unit tests

guilpier-code committed:
add solver options (optim 1 and 2) : typo

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

pet-mit committed:
fixes after merge

pet-mit committed:
Merge branch 'develop' into feature/new_quad_solvers

pet-mit committed:
clang

guilpier-code committed:
[skip ci] add solver options (optim 1 and 2) : solver menu (--help) enhancement + doc

antares-solver menu had flaws in columns alignement.
Documentation is supplied.

guilpier-code committed:
add solver options (optim 1 and 2) : simplifying the code a bit

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/feature/new_quad_solvers' into feature/add-solver-options-for-optim-1-and-2

guilpier-code committed:
add solver options (optim 1 and 2) : correction

guilpier-code committed:
[skip ci] add solver options (optim 1 and 2) : clang format

flomnes commented on src/solver/misc/options.cpp:
```diff
@@ -82,45 +79,53 @@ std::unique_ptr<Yuni::GetOpt::Parser> CreateParser(Settings& settings, StudyLoad
+    // --lp-solver-param-optim-1
+    parser->add(options.solverOptions.lpSolverParamOptim1,
+                ' ',
+                "lp-solver-param-optim-1",
```
```suggestion
                "linear-solver-parameters-optim1",
```

guilpier-code replied:
See below

flomnes commented on src/solver/misc/options.cpp:
```diff
@@ -82,45 +79,53 @@ std::unique_ptr<Yuni::GetOpt::Parser> CreateParser(Settings& settings, StudyLoad
+    // --lp-solver-param-optim-2
+    parser->add(options.solverOptions.lpSolverParamOptim2,
+                ' ',
+                "lp-solver-param-optim-2",
```
```suggestion
                "linear-solver-parameters-optim2",
```

guilpier-code replied:
We finally opted for **--lp-solver-param**, **--lp-solver-param-optim-1** and **--lp-solver-param-optim-2**

Please see this[ overall PR ](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/2765)
